### PR TITLE
Add TorchScriptWrapper_v1

### PR DESCRIPTION
**Pull request by danieldk**
PyTorch `Module`s can be converted to TorchScript. TorchScript has the advantage that the model is serialized with the model parameters in a portable manner. Deserialization (in contrast to pickling) does not require certain Python types to be available. In fact, a TorchScript module can be loaded in C++ without a Python interpreter.

For Thinc and spaCy, supporting TorchScript has two main benefits:

1. Since the model is serialized, we don't run into the issue that we have with the PyTorch wrapper where we have to construct the PyTorch model before deserializing its parameters but we can only know the model shapes through deserialization.

2. When a model is rewritten for e.g. quantization, it is unwieldy to construct the rewritten model by hand. So, with the TorchScript wrapper we would first have to construct the original model and then reapply the graph transformation. This is quite bad for transformations that reduce the model size, since we end up temporarily allocating all parameters. This is not an issue with TorchScript, since we serialize the rewritten model.

### CODE REVIEW

**Review by shadeMe — APPROVED** [review]

**Comment by svlandeg** [comment]
@danieldk : can you have a look at the conflicts?

**Comment by danieldk** [comment]
> @danieldk : can you have a look at the conflicts?

Fixed.

**Review by svlandeg — COMMENTED** [review]
This will be great to have! I'm particularly enthousiastic about the prospect of resolving the initialization/deserialization shape issues 🎉

File: thinc/shims/torchscript.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        in half precision for better performance and lower memory use.
+    grad_scaler:
+        The gradient scaler to use for mixed-precision training. If this
+        argument is set to "None" and mixed precision is enabled, a gradient
+        scaler with the default configuration is used.
+    device:
+        The PyTorch device to run the model on. When this argument is
+        set to "None", the default device for the currently active Thinc
+        ops is used.
+    """
```
[the remark this answers is not in the document]
Comment (danieldk) [reply thinc/shims/torchscript.py:28]: I am not really sure how to document them. We never actually use the `optimizer` argument. The `config` argument is  serialized/deserialized as `self.cfg`, but also not used by any shim?

File: thinc/layers/torchscriptwrapper.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
+from typing import Any, Callable, Optional
+from thinc.model import Model
```
[the remark this answers is not in the document]
Comment (danieldk) [reply thinc/layers/torchscriptwrapper.py]: Corrected.

File: thinc/layers/torchscriptwrapper.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    mixed_precision: bool = False,
+    grad_scaler: Optional[PyTorchGradScaler] = None,
+    device: Optional["torch.device"] = None,
+) -> Model[Any, Any]:
+    """Wrap a TorchScript model, so that it has the same API as Thinc models.
+
+    model:
+        The TorchScript module. A value of `None` is also possible to
+        construct a shim to deserialize into.
+    mixed_precision:
```
[the remark this answers is not in the document]
Comment (danieldk) [reply thinc/layers/torchscriptwrapper.py:29]: Added.

File: thinc/layers/torchscriptwrapper.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    torchscript_model: Optional["torch.jit.ScriptModule"] = None,
+    convert_inputs: Optional[Callable] = None,
+    convert_outputs: Optional[Callable] = None,
+    mixed_precision: bool = False,
+    grad_scaler: Optional[PyTorchGradScaler] = None,
+    device: Optional["torch.device"] = None,
+) -> Model[Any, Any]:
+    """Wrap a TorchScript model, so that it has the same API as Thinc models.
+
+    model:
```
[the remark this answers is not in the document]
Comment (danieldk) [reply thinc/layers/torchscriptwrapper.py]: Fixed.

File: thinc/layers/torchscriptwrapper.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+def pytorch_to_torchscript_wrapper(model: Model):
+    """Convert a PyTorch wrapper to a TorchScript wrapper. The embedded PyTorch
+    `Module` is converted to `ScriptModule`.
+    """
+
+    if model.name != "pytorch":
+        raise ValueError(
+            "Only PyTorch wrappers can be converted to TorchScript wrappers"
+        )
```
[the remark this answers is not in the document]
Comment (danieldk) [reply thinc/layers/torchscriptwrapper.py]: Yeah, agreed. I removed this check.

File: thinc/shims/__init__.py [hunk 6]
```diff
@@ -12,5 +13,6 @@
     "PyTorchGradScaler",
     "keras_model_fns", "TensorFlowShim", "maybe_handshake_model",
```
[the remark this answers is not in the document]
Comment (danieldk) [reply thinc/shims/__init__.py]: Done, and also sorted alphabetically.

File: thinc/shims/torchscript.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    def __init__(
+        self,
+        model: Optional["torch.ScriptModule"],
+        config=None,
+        optimizer: Any = None,
+        mixed_precision: bool = False,
+        grad_scaler: Optional[PyTorchGradScaler] = None,
+        device: Optional["torch.device"] = None,
+    ):
+        if not isinstance(model, torch.jit.ScriptModule) and model is not None:
```
[the remark this answers is not in the document]
Comment (danieldk) [reply thinc/shims/torchscript.py]: 💯  Fixed.

File: website/docs/api-layers.md [hunk 8]
```diff
@@ -1695,6 +1695,63 @@ the backward pass.
...
+model = PyTorchWrapper_v2(torch.nn.Linear(nI, nO)).initialize()
+script_model = pytorch_to_torchscript_wrapper(model)
+```
+
+| Argument            | Type                                      | Description                                                                              |
+| ------------------- | ----------------------------------------- | ---------------------------------------------------------------------------------------- |
+| `torchscript_model` | <tt>Optional[torch.jit.ScriptModule]</tt> | The TorchScript model.                                                                   |
+| `convert_inputs`    | <tt>Callable</tt>                         | Function to convert inputs to PyTorch tensors (same signature as `forward` function).    |
+| `convert_outputs`   | <tt>Callable</tt>                         | Function to convert outputs from PyTorch tensors (same signature as `forward` function). |
+| **RETURNS**         | <tt>Model[Any, Any]</tt>                  | The Thinc model.                                                                         |
```
[the remark this answers is not in the document]
Comment (danieldk) [reply website/docs/api-layers.md]: Added.

File: website/docs/usage-frameworks.md [hunk 9]
```diff
@@ -37,11 +37,12 @@ via the [DLPack](https://github.com/dmlc/dlpack) standard wherever possible, so
...
+| -------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------ | ------------------ |
+| **PyTorch**    | [`PyTorchWrapper`](/docs/api-layers#pytorchwrapper) ([code](https://github.com/explosion/thinc/blob/master/thinc/layers/pytorchwrapper.py))             | [`PyTorchShim`](/docs/api-model#shims) ([code](https://github.com/explosion/thinc/blob/master/thinc/shims/pytorch.py))         | <i name="yes"></i> |
+| **PyTorch**    | [`TorchScriptWrapper`](/docs/api-layers#torchscriptwrapper) ([code](https://github.com/explosion/thinc/blob/master/thinc/layers/torchscriptwrapper.py)) | [`TorchScriptShim`](/docs/api-model#shims) ([code](https://github.com/explosion/thinc/blob/master/thinc/shims/torchscript.py)) | <i name="yes"></i> |
```
[the remark this answers is not in the document]
Comment (danieldk) [reply website/docs/usage-frameworks.md]: Fixed.

**Comment by danieldk** [comment]
I put the PR in draft for the moment, since I am still wondering if it's best to pass `None` as defaults to the conversion functions.

File: thinc/shims/torchscript.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (svlandeg) [reply thinc/shims/torchscript.py:28]: Meh ok. Let's leave it ;-)

**Comment by svlandeg** [comment]
> I put the PR in draft for the moment, since I am still wondering if it's best to pass `None` as defaults to the conversion functions.

As discussed, I'd keep it as is.
